### ISSUE TITLE
Crash in contract_executor_t on raft branch

### ISSUE THREAD

**Issue by timmaxw** [rethinkdb/rethinkdb#4254]
```
error: Error in ../src/concurrency/mutex_assertion.hpp at line 32:
error: Assertion failed: [!mutex->locked] 
error: Backtrace:
error: Wed May 20 20:43:16 2015
       
       1: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:100
       2: backtrace_t::backtrace_t() at backtrace.cc:203
       3: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
       4: format_backtrace(bool) at backtrace.cc:198
       5: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
       6: mutex_assertion_t::acq_t::reset(mutex_assertion_t*) at mutex_assertion.hpp:32
       7: mutex_assertion_t::acq_t::acq_t(mutex_assertion_t*) at mutex_assertion.hpp:19
       8: mutex_assertion_t::acq_t::acq_t(mutex_assertion_t*) at mutex_assertion.hpp:19
       9: pump_coro_t::notify() at pump_coro.cc:22
       10: contract_executor_t::update_blocking(signal_t*) at executor.cc:115
       11: std::_Mem_fn<void (contract_executor_t::*)(signal_t*)>::operator()(contract_executor_t*, signal_t*) const at functional:551
       12: void std::_Bind<std::_Mem_fn<void (contract_executor_t::*)(signal_t*)> (contract_executor_t*, std::_Placeholder<1>)>::__call<void, signal_t*&&, 0, 1>(std::tuple<signal_t*&&>&&, std::_Index_tuple<0, 1>) at functional:1146
       13: void std::_Bind<std::_Mem_fn<void (contract_executor_t::*)(signal_t*)> (contract_executor_t*, std::_Placeholder<1>)>::operator()<signal_t*, void>(signal_t*&&) at functional:1204
       14: std::_Function_handler<void (signal_t*), std::_Bind<std::_Mem_fn<void (contract_executor_t::*)(signal_t*)> (contract_executor_t*, std::_Placeholder<1>)> >::_M_invoke(std::_Any_data const&, signal_t*) at functional:1780
       15: std::function<void (signal_t*)>::operator()(signal_t*) const at functional:2161
       16: pump_coro_t::run(auto_drainer_t::lock_t) at pump_coro.cc:88
       17: std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)>::operator()(pump_coro_t*, auto_drainer_t::lock_t) const at functional:551
       18: void std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)>::__call<void, , 0, 1>(std::tuple<>&&, std::_Index_tuple<0, 1>) at functional:1146
       19: void std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)>::operator()<, void>() at functional:1204
       20: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
       21: callable_action_wrapper_t::run() at runtime_utils.cc:43
       22: coro_t::run() at coroutines.cc:205
       23: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (pump_coro_t::*)(auto_drainer_t::lock_t)> (pump_coro_t*, auto_drainer_t::lock_t)>&&) at coroutines.hpp:58
       24: pump_coro_t::notify() at pump_coro.cc:31
       25: /home/tim/rethinkdb/build/debug_clang/rethinkdb() [0x1e2d5a7] at 0x1e2d5a7 ()
       26: /home/tim/rethinkdb/build/debug_clang/rethinkdb() [0x1e2d2cd] at 0x1e2d2cd ()
       27: std::function<void ()>::operator()() const at functional:2161
       28: call_function(std::function<void ()> const&) at watchable.hpp:185
       29: void publisher_controller_t<std::function<void ()> >::publish<void (*)(std::function<void ()> const&)>(void (* const&)(std::function<void ()> const&)) at pubsub.hpp:139
       30: /home/tim/rethinkdb/build/debug_clang/rethinkdb() [0x1d4695a] at 0x1d4695a ()
       31: /home/tim/rethinkdb/build/debug_clang/rethinkdb() [0x1d47e42] at 0x1d47e42 ()
       32: /home/tim/rethinkdb/build/debug_clang/rethinkdb() [0x1d47dc3] at 0x1d47dc3 ()
       33: /home/tim/rethinkdb/build/debug_clang/rethinkdb() [0x1d47d76] at 0x1d47d76 ()
       34: /home/tim/rethinkdb/build/debug_clang/rethinkdb() [0x1d4793d] at 0x1d4793d ()
       35: std::function<void ()>::operator()() const at functional:2161
       36: call_function(std::function<void ()> const&) at watchable.hpp:185
       37: void publisher_controller_t<std::function<void ()> >::publish<void (*)(std::function<void ()> const&)>(void (* const&)(std::function<void ()> const&)) at pubsub.hpp:139
       38: watchable_variable_t<raft_member_t<table_raft_state_t>::state_and_config_t>::apply_atomic_op(std::function<bool (raft_member_t<table_raft_state_t>::state_and_config_t*)> const&) at watchable.hpp:223
       39: raft_member_t<table_raft_state_t>::update_commit_index(unsigned long, new_mutex_acq_t const*) at raft_core.tcc:1061
       40: raft_member_t<table_raft_state_t>::leader_update_match_index(raft_member_id_t, unsigned long, new_mutex_acq_t const*, signal_t*) at raft_core.tcc:1142
       41: raft_member_t<table_raft_state_t>::leader_send_updates(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t) at raft_core.tcc:1794
       42: std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)>::operator()(raft_member_t<table_raft_state_t>*, raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t) const at functional:551
       43: void std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, raft_member_id_t, unsigned long, auto_drainer_t::lock_t)>::__call<void, , 0, 1, 2, 3>(std::tuple<>&&, std::_Index_tuple<0, 1, 2, 3>) at functional:1146
       44: void std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, raft_member_id_t, unsigned long, auto_drainer_t::lock_t)>::operator()<, void>() at functional:1204
       45: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, raft_member_id_t, unsigned long, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
       46: callable_action_wrapper_t::run() at runtime_utils.cc:43
       47: coro_t::run() at coroutines.cc:205
       48: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, raft_member_id_t, unsigned long, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, raft_member_id_t, unsigned long, auto_drainer_t::lock_t)>&&) at coroutines.hpp:58
       49: raft_member_t<table_raft_state_t>::leader_spawn_update_coros(unsigned long, std::map<raft_member_id_t, scoped_ptr_t<auto_drainer_t>, std::less<raft_member_id_t>, std::allocator<std::pair<raft_member_id_t const, scoped_ptr_t<auto_drainer_t> > > >*, new_mutex_acq_t const*) at raft_core.tcc:1575
       50: raft_member_t<table_raft_state_t>::candidate_and_leader_coro(new_mutex_acq_t*, auto_drainer_t::lock_t) at raft_core.tcc:1325
       51: std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)>::operator()(raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t) const at functional:551
       52: void std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t)>::__call<void, , 0, 1, 2>(std::tuple<>&&, std::_Index_tuple<0, 1, 2>) at functional:1146
       53: void std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t)>::operator()<, void>() at functional:1204
       54: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
       55: callable_action_wrapper_t::run() at runtime_utils.cc:43
       56: coro_t::run() at coroutines.cc:205
       57: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t)>&&) at coroutines.hpp:58
       58: raft_member_t<table_raft_state_t>::on_watchdog_timer()::{lambda()#1}::operator()() const at raft_core.tcc:991
       59: callable_action_instance_t<raft_member_t<table_raft_state_t>::on_watchdog_timer()::{lambda()#1}>::run_action() at callable_action.hpp:31
       60: callable_action_wrapper_t::run() at runtime_utils.cc:43
       61: coro_t::run() at coroutines.cc:205
       62: coro_t* coro_t::spawn_sometime<raft_member_t<table_raft_state_t>::on_watchdog_timer()::{lambda()#1}>(raft_member_t<table_raft_state_t>::on_watchdog_timer()::{lambda()#1}&&) at coroutines.hpp:58
       63: raft_member_t<table_raft_state_t>::on_watchdog_timer() at raft_core.tcc:969
       64: raft_member_t<table_raft_state_t>::raft_member_t(raft_member_id_t const&, raft_storage_interface_t<table_raft_state_t>*, raft_network_interface_t<table_raft_state_t>*, raft_persistent_state_t<table_raft_state_t> const&, std::string const&)::{lambda()#1}::operator()() const at raft_core.tcc:92
       65: std::_Function_handler<void (), raft_member_t<table_raft_state_t>::raft_member_t(raft_member_id_t const&, raft_storage_interface_t<table_raft_state_t>*, raft_network_interface_t<table_raft_state_t>*, raft_persistent_state_t<table_raft_state_t> const&, std::string const&)::{lambda()#1}>::_M_invoke(std::_Any_data const&) at functional:1780
       66: std::function<void ()>::operator()() const at functional:2161
       67: call_ringer(std::function<void ()>) at timing.cc:93
       68: void std::_Bind<void (*(std::function<void ()>))(std::function<void ()>)>::__call<void, , 0>(std::tuple<>&&, std::_Index_tuple<0>) at functional:1146
       69: void std::_Bind<void (*(std::function<void ()>))(std::function<void ()>)>::operator()<, void>() at functional:1204
       70: callable_action_instance_t<std::_Bind<void (*(std::function<void ()>))(std::function<void ()>)> >::run_action() at callable_action.hpp:31
       71: callable_action_wrapper_t::run() at runtime_utils.cc:43
       72: coro_t::run() at coroutines.cc:205
       73: void coro_t::spawn_now_dangerously<std::_Bind<void (*(std::function<void ()>))(std::function<void ()>)> >(std::_Bind<void (*(std::function<void ()>))(std::function<void ()>)>&&) at coroutines.hpp:52
       74: repeating_timer_t::on_timer() at timing.cc:99
       75: timer_handler_t::on_oneshot() at timer.cc:62
       76: timerfd_provider_t::on_event(int) at timerfd_provider.cc:86
       77: epoll_event_queue_t::run() at epoll.cc:115
       78: linux_thread_pool_t::start_thread(void*) at thread_pool.cc:155
       79: /lib/x86_64-linux-gnu/libpthread.so.0(+0x8182) [0x7f8e485d2182] at 0x7f8e485d2182 (/lib/x86_64-linux-gnu/libpthread.so.0)
       80: clone+0x6d at 0x7f8e482fefbd (/lib/x86_64-linux-gnu/libc.so.6)
error: Exiting.
```

Observed when running the `shard_fuzzer.py` test on a70e13798f56435bc5c792470760e7ec21da7400.

**Comment by Tryneus** [comment]
I suspect this is due to the addition of `pump_coro_t::drain()` - if someone tries to notify the `pump_coro_t` while it is still draining, the `mutex_assertion_t::acq_t` will fail.  It might fix it if we check for `drained` before acquiring the `mutex_assertion_t`.  Thoughts, @timmaxw?

**Comment by timmaxw** [comment]
I agree with your diagnosis, but I think we should fix it by not holding the mutex assertion while we're draining.

**Comment by Tryneus** [comment]
I'll put this up for review then.

**Comment by Tryneus** [comment]
The fix has been merged to `raft` in commit 12940fa, review was 2915.
